langworld_db_data/adders/listed_value_adder.py

```diff
@@ -59,11 +84,47 @@ def _add_to_inventory_of_listed_values(
                     f"Row {row} already contains value you are trying to add"
                 )
 
-            # Keep updating those with each row. This means that at the last row of the
-            # feature they will reach the required values.
-            index_of_last_row_for_given_feature = i
             last_digit_of_value_id = int(row["id"].split(ID_SEPARATOR)[-1])
-            id_of_new_value = feature_id + ID_SEPARATOR + str(last_digit_of_value_id + 1)
+            values_diapason.append(
+                {
+                    "index": last_digit_of_value_id,
+                    "row": i,
+                }
+            )
+
+        last_index_in_feature = values_diapason[-1]["index"]
+        if index_to_assign > last_index_in_feature:
+            raise ListedValueAdderError(
+                f"ID {feature_id + str(index_to_assign)} exceeds the maximal ID {last_index_in_feature}"
+            )
+
+        id_of_new_value = f"{feature_id}{ID_SEPARATOR}{values_diapason[-1]['index'] + 1}"
+        row_of_new_value = values_diapason[-1]["row"] + 1
+        rows_with_ids_to_increment = []
+
+        if index_to_assign > -1:
+            id_of_new_value = f"{feature_id}{ID_SEPARATOR}{index_to_assign}"
+            for index_row in values_diapason:
+                if index_row["index"] < index_to_assign:
+                    continue
+
+                rows_with_ids_to_increment.append(index_row["row"])
+                if index_row["index"] == index_to_assign:
+                    row_of_new_value = index_row["row"]
```

Comment:
my intuition says that you should break out of the loop after this assignment?

Reply:
This part of the loop may be misleading. The loop collects indices that must be incremented in features_listed_values and at the same time it finds the line number for the new value. Perhaps these should be two separate loops.

Also, while writing comments in code, I found that two variables, `line_numbers_with_ids_to_increment` (former `rows_with_ids_to_increment`) and `ids_to_increment_in_profiles`, are virtually identical and that it is possible to rewrite this code segment in a briefer way, without both of them. In this case making two separate loops may be even more favourable, to my mind.

Reply:
> Also, while writing comments in code, I found that two variables, `line_numbers_with_ids_to_increment` (former `rows_with_ids_to_increment`) and `ids_to_increment_in_profiles`, are virtually identical and that it is possible to rewrite this code segment in a briefer way, without the first one. In this case making two separate loops may be even more favourable, to my mind.

That's why it's good to comment your own code :) If you found a way to use just one variable and make code briefer and more readable, you certainly should do it :)

Reply:
Okay! The code is now rewritten with only one variable and moved to a separate method (which is also provided with its own test)